README.md

```diff
@@ -1,12 +1,332 @@
+
+$$\boldsymbol A^{-1} \boldsymbol b.$$
+
+If you didn't know anything about the matrix $\boldsymbol A$, the simplest (and best) way to accomplish this in code is:
+
+```python
+# A = torch.randn(1000, 1000)
+# b = torch.randn(1000)
+torch.linalg.solve(A, b)  # computes A^{-1} b
+```
+
+While this is easy, the `solve` routine is $\mathcal O(N^3)$, which gets very slow as $N$ grows large.
+
+However, let's imagine that we knew that $\boldsymbol A$ was equal to a low rank matrix plus a diagonal
+(i.e. $\boldsymbol A = \boldsymbol C \boldsymbol C^\top + \boldsymbol D$
+for some skinny matrix $\boldsymbol C$ and some diagonal matrix $\boldsymbol D$.)
+There's now a very efficient $\boldsymbol O(N)$ routine to compute $\boldsymbol A^{-1}$
+called the [Woodbury formula](https://en.wikipedia.org/wiki/Woodbury_matrix_identity).
+**In general**, if we know that $\boldsymbol A$ has structure,
+we want to use efficient linear algebra routines (rather than the general routines)
+that will exploit this structure.
+
+
+### Without LinearOperator
+
+Implementing the efficient solve that exploits the structure in $\boldsymbol A$ would look something like this:
+
+```python
+def low_rank_plus_diagonal_solve(C, d, b):
+    # A = C C^T + diag(d)
+    # A^{-1} b = D^{-1} b - D^{-1} C (I + C^T D^{-1} C)^{-1} C^T D^{-1} b
+    #   where D = diag(d)
+
+    D_inv_b = b / d
+    D_inv_C = C / d.unsqueeze(-1)
+    eye = torch.eye(C.size(-2))
+    return (
+        D_inv_b - D_inv_c @ torch.cholesky_solve(
+            C.mT @ D_inv_b,
+            torch.linalg.cholesky(eye + C.mT @ D_invC, upper=False),
+            upper=False
+        )
+    )
+
+
+# C = torch.randn(1000, 20)
+# d = torch.randn(1000)
+# b = torch.randn(1000)
+low_rank_plus_diagonal_solve(C, d, b)  # computes A^{-1} b efficiently
+```
+
+While this is efficient code, it's not ideal for a number of reasons:
+1. It's a lot more complicated than `torch.linalg.solve(A, b)`.
+2. There's no object that represents $\boldsymbol A$.
+   To perform any math with $\boldsymbol A$, we have to pass around the matrix `C` and the vector `d`.
+
+
+### With LinearOperator
+
+The LinearOperator package offers the best of both worlds:
+
+```python
+from linear_operator.operators import DiagLinearOperator, RootLinearOperator
+# C = torch.randn(1000, 20)
+# d = torch.randn(1000)
+# b = torch.randn(1000)
+A = RootLinearOperator(C) + DiagLinearOperator(d)  # represents C C^T + diag(d)
+```
+
+it provides an interface that lets us treat $\boldsymbol A$ as if it were a generic tensor,
+using the standard PyTorch API...
+
+```python
+torch.linalg.solve(A, d, b)  # computes A^{-1} b efficiently!
 ```
-python -m unittest discover
+
+but under-the-hood, the `LinearOperator` object keeps track of the algebraic structure of $\boldsymbol A$ (low rank plus diagonal)
+and determines the most efficient routine to use (the Woodbury formula).
+This way, we can get a efficient ($\mathcal O(N)$) solve while abstracting away all of the details.
```

Comment:
this doesn't seem to render properly, probably b/c of the outer parentheses?

Reply:
Fixed